File: leetcode/best_time_to_buy_and_sell_stock.c
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

#define min(x,y) ((y) + (((x) - (y)) & (((x) - (y)) >> (sizeof(int) * CHAR_BIT - 1))))
#define max(x,y) ((x) - (((x) - (y)) & (((x) - (y)) >> (sizeof(int) * CHAR_BIT - 1))))

int maxProfit(int prices[], int n) {
    int max_profit = 0;
    int i, curr_min; 
    if (n >= 2) {
        curr_min = prices[0];
        for (i = 1; i < n; i++) {
            curr_min = min(curr_min, prices[i]);
            max_profit = max(max_profit, prices[i] - curr_min);
        }
    }
    return max_profit;
}

File: leetcode/best_time_to_buy_and_sell_stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        max_profit = 0
        if (len(prices) > 0):
            curr_min = prices[0]
            for price in prices[1:]:
                curr_min = min(curr_min, price) 
                max_profit = max(max_profit, price - curr_min)
        return max_profit